Add documentation for Pulumi registry (#27)
1Password/pulumi-onepassword: docs/installation-configuration.md

---
title: 1Password: Installation & Configuration

> meta_desc: Provides an overview on how to configure credentials for the 1Password Pulumi Provider.

Reviewer: Are we using "1Password Pulumi Provider" as a product name? If so, we should capitalize "Provider" elsewhere in the docs. If not, we should lowercase "provider" here.

Author: I don't think we have a product name for this per se. I think `1Password Provider for Pulumi` or `1Password Pulumi Provider` are common to use for any provider for Pulumi. 

Looking through other documentation in [Pulumi registry[(https://www.pulumi.com/registry/) I see that the patter `X provider for Pulumi` is more commonly used. so I'm leaning towards changing this everywhere in this way for consistency.

layout: package
---

> The 1Password Pulumi provider enables accessing and managing items in your 1Password vaults.

Reviewer: ```suggestion
The 1Password Pulumi provider allows you to access and manage items in your 1Password vaults.
```

To follow our [style rule on "enable"](https://support.1password.com/style-guide/#enable--disable)

Also, should we mention Pulumi here?

```suggestion
The 1Password Pulumi provider allows you to access and manage items in your 1Password vaults using Pulumi.
```

> ## Installation

Reviewer: Do these headers follow a standardized template from Pulumi, or can we make changes to them?

If we can change them:

```suggestion
## Install the 1Password Pulumi provider
```

Author: It looks like it's quite common to use `Installation` as a header, so I'm leaning towards keeping it as is to be consistent.

The 1Password Pulumi provider is available as a package in most Pulumi languages:

- JavaScript/TypeScript: [@1password/pulumi-onepassword](https://www.npmjs.com/package/@1password/pulumi-onepassword)
- Python: [pulumi_onepassword](https://pypi.org/project/pulumi-onepassword/)
- Go: [github.com/1Password/pulumi-onepassword/sdk/go/onepassword](https://pkg.go.dev/github.com/1Password/pulumi-onepassword/sdk/go/onepassword)
- .NET: _coming soon_

> ## Configuring Credentials

Reviewer: ```suggestion
## Configure your 1Password credentials
```

Author: How about simply `Configuration`? I see other providers having this heading and it might be easier to read and identify. 🤔

> The 1Password pulumi provider needs to be configured with 1Password credentials before it can be used to access and manage items in your 1Password vault. You can configure it to use one of the following:

Reviewer: ```suggestion
You must configure the 1Password Pulumi provider with your 1Password credentials before the provider can be used to access and manage items in your 1Password vaults. You can use one of the following:
```

To avoid passive voice

> - Service account
>   - `pulumi-onepassword:service_account_token` (environment: `OP_SERVICE_ACCOUNT_TOKEN`) - The 1Password service account token to use with 1Password CLI.
> - Connect
>   - `pulumi-onepassword:url` (environment: `OP_CONNECT_HOST`) - the URL where your 1Password Connect API can be found.
>   - `pulumi-onepassword:token` (environment: `OP_CONNECT_TOKEN`) - the token for your Connect API.
> - Personal account
>   - `pulumi-onepassword:account` (environment: `OP_ACCOUNT`) - A valid account's sign-in address or ID to use with 1Password CLI and biometric unlock.

Reviewer: ```suggestion
- [Service account](https://developer.1password.com/docs/service-accounts/get-started)
  - `pulumi-onepassword:service_account_token` (environment: `OP_SERVICE_ACCOUNT_TOKEN`) - the 1Password service account token to use with 1Password CLI.
- [Connect server](https://developer.1password.com/docs/connect/get-started)
  - `pulumi-onepassword:url` (environment: `OP_CONNECT_HOST`) - the URL where your 1Password Connect API can be found.
  - `pulumi-onepassword:token` (environment: `OP_CONNECT_TOKEN`) - the token for your Connect API.
- Personal account
  - `pulumi-onepassword:account` (environment: `OP_ACCOUNT`) - the [sign-in address](https://support.1password.com/1password-glossary/#sign-in-address) or [unique identifier](https://developer.1password.com/docs/cli/reference/#unique-identifiers-ids) for your 1Password account to use with 1Password CLI and biometric unlock.
```

Edits to add links to documentation and some small style changes. 

Also, a question: why do we mention 1Password CLI here? Do they need to have the CLI installed to use those options?

Author: For the case of service account and personal account, the 1Password CLI is required.

Reviewer: Should we mention that here? Maybe instead of saying "to use with 1Password CLI" we could say something like "Requires 1Password CLI". Or "Uses 1Password CLI to authenticate". Or something similar that indicates it's a requirement, then link out to the CLI get started docs.

Author: I feel like it's a good idea to do so. I've chosen to go with `Requires 1Password CLI` since it was short and on point.

> Once the credentials obtained, there are two ways to communicate your configuration to Pulumi:

Reviewer: ```suggestion
After you find your credentials, there are two ways to provide them to Pulumi:
```

> 1. Set the environment variables for the preferred method:

Reviewer: ```suggestion
1. Set the environment variables for the preferred method. For example, to set the environment variable for a service account:
```

   ```sh
   $ export OP_SERVICE_ACCOUNT_TOKEN=XXXXXXXXXXXXXX
   ```

> 2. Set them using configuration, if you prefer that they be stored alongside your Pulumi stack for easy multi-user access:

Reviewer: ```suggestion
2. If you prefer to store your credentials alongside your Pulumi stack for easy multi-user access, use configuration to set them.
```

To avoid passive voice and lead with the value proposition here

>   - `pulumi-onepassword:account` (environment: `OP_ACCOUNT`) - the [sign-in address](https://support.1password.com/1password-glossary/#sign-in-address) or [unique identifier](https://developer.1password.com/docs/cli/reference/#unique-identifiers-ids) for your 1Password account to use. Requires [1Password CLI](https://developer.1password.com/docs/cli/get-started/) and biometric unlock.

Reviewer: ```suggestion
  - `pulumi-onepassword:account` (environment: `OP_ACCOUNT`) - the [sign-in address](https://support.1password.com/1password-glossary/#sign-in-address) or [unique identifier](https://developer.1password.com/docs/cli/reference/#unique-identifiers-ids) for your 1Password account. Requires [1Password CLI](https://developer.1password.com/docs/cli/get-started/) and biometric unlock.
```

   ```sh
   $ pulumi config set pulumi-onepassword:service_account_token XXXXXXXXXXXXXX --secret
   ```

> Remember to pass `--secret` when setting any sensitive data (in this example `pulumi-onepassword:service_account_token`) so that it is properly encrypted. The complete list of configuration parameters is in the [1Password Pulumi provider README](https://github.com/1Password/pulumi-onepassword/blob/main/README.md).

Reviewer: ```suggestion
Make sure to pass `--secret` when setting any sensitive data (in this example `pulumi-onepassword:service_account_token`) so that it's properly encrypted. The complete list of configuration parameters is in the [1Password Pulumi provider README](https://github.com/1Password/pulumi-onepassword/blob/main/README.md).
```

Also, is it possible to tab this so it appears under option 2? If not, could we use subheaders instead of a numbered list for the options?

i.e.

### Option 1: Using environment variables

### Option 2: Using configuration

Author: I believe it is possible, although I am not 100% sure. We might need to as the Pulumi folks about this.

Reviewer: Okay, not a big deal if it's not!

Author: Based on how the tabs seems to be made when doing code snippets, I'd assume having tabs here would be possible with a code snippet like this:

```
{{< chooser "env-var,config" >}}

{{% choosable env-var %}}

// documentation for env-var

{{% /choosable %}}
{{% choosable config %}}

// documentation for pulumi config

{{% /choosable %}}
{{< /chooser >}}
```

Author: Closing this for now to not block the docs from being published, but feel free to share your opinion on this even after it's merged.

Reviewer 2: @edif2008 the `chooser`/`choosable` setup should indeed work. The tags need 2 arguments though: a type and value(s). Example for your configuration:

```
{{< chooser configuration "env-var,pulumi-config" >}}

{{% choosable configuration env-var %}}

// documentation for env-var

{{% /choosable %}}
{{% choosable configuration pulumi-config %}}

// documentation for pulumi config

{{% /choosable %}}
{{< /chooser >}}
```

Using this search query, you can see how we use it ourselves for types `language` and `os` with a different set of values. Hope this helps.
